pom.xml

```diff
@@ -67,7 +67,7 @@
   <org.apache.commons.validator.version>1.7</org.apache.commons.validator.version>
   <org.apache.commons.io.version>2.8.0</org.apache.commons.io.version>
   <org.apache.httpcomponents.version>4.5.13</org.apache.httpcomponents.version>
-  <io.fabric8.client.version>5.3.0</io.fabric8.client.version>
+  <io.fabric8.client.version>5.6.0</io.fabric8.client.version>
```

Comment:
Does 5.4.1 contain the fix you needed?

Reply:
I think so - I believe this was the issue for the specific bug I encountered: https://github.com/fabric8io/kubernetes-client/issues/2529

That's labelled with the 5.4.0 milestone, so I'll test out 5.4.1 but I expect it should do it. I suppose you ask because that version is already available for downstream?

Reply:
Right. Should make things a bit easier for us.

Reply:
https://github.com/fabric8io/kubernetes-client/issues/2265

https://github.com/fabric8io/kubernetes-client/issues/2264

are also relevant.

When I change the fabric8 client version in the pom.xml to 5.4.1, the server-mock unit tests I added in the last couple of commits fail. They pass with 5.3.0, 5.3.1, 5.5.0, and 5.6.0 however. I haven't found any bug reports that indicate why this might be however, but because of the discrepancy between versions running the same unit tests I'm suspecting it might be another bug in the resource implementation in the kubernetes-client.

The unit test failure in 5.4.1 looks like this:


```
[ERROR] shouldValidateTokenWithSufficientPermissions  Time elapsed: 2.445 s  <<< ERROR!
java.util.concurrent.ExecutionException: java.util.concurrent.ExecutionException: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: http://localhost:37341/apis/authorization.k8s.io/v1/selfsubjectacce
ssreviews. Message: SelfSubjectAccessReview is invalid. Received status: Status(apiVersion=v1, code=422, details=StatusDetails(causes=[StatusCause(field=null, message=Required value: metadata is required, reason=ValueRequired, additi
onalProperties={})], group=null, kind=SelfSubjectAccessReview, name=null, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=SelfSubjectAccessReview is invalid, metadata=null, reason=Invalid, status=Fail
ure, additionalProperties={}).
        at io.cryostat.net.OpenShiftAuthManagerTest.shouldValidateTokenWithSufficientPermissions(OpenShiftAuthManagerTest.java:142)
Caused by: java.util.concurrent.ExecutionException: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: http://localhost:37341/apis/authorization.k8s.io/v1/selfsubjectaccessreviews. Message: SelfSubjec
tAccessReview is invalid. Received status: Status(apiVersion=v1, code=422, details=StatusDetails(causes=[StatusCause(field=null, message=Required value: metadata is required, reason=ValueRequired, additionalProperties={})], group=nul
l, kind=SelfSubjectAccessReview, name=null, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=SelfSubjectAccessReview is invalid, metadata=null, reason=Invalid, status=Failure, additionalProperties={}).
        at io.cryostat.net.OpenShiftAuthManagerTest.shouldValidateTokenWithSufficientPermissions(OpenShiftAuthManagerTest.java:142)
Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: http://localhost:37341/apis/authorization.k8s.io/v1/selfsubjectaccessreviews. Message: SelfSubjectAccessReview is invalid. Received status
: Status(apiVersion=v1, code=422, details=StatusDetails(causes=[StatusCause(field=null, message=Required value: metadata is required, reason=ValueRequired, additionalProperties={})], group=null, kind=SelfSubjectAccessReview, name=nul
l, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=SelfSubjectAccessReview is invalid, metadata=null, reason=Invalid, status=Failure, additionalProperties={}).
        at io.cryostat.net.OpenShiftAuthManagerTest.shouldValidateTokenWithSufficientPermissions(OpenShiftAuthManagerTest.java:142)
```

This is very similar to the failed results I saw when running with 5.3.1 against an actual CRC cluster. So, I wonder if 5.4.0/5.4.1 fixed the implementation in kubernetes-client but didn't fix the behaviour of the kubernetes-server-mock, or something along those lines...

Reply:
Maybe the SelfSubjectAccessReview needs to be given a name?

Reply:
Tried that just now, no luck.

Reply:
I figured it out. There is an OkHttp Interceptor added in the DefaultOpenShiftClient to retry requests that fail authorization by finding a token or asking the server for a new one:
https://github.com/fabric8io/kubernetes-client/blob/7e747ca61c0097562622eb53b9f008cc6d5e82fb/openshift-client/src/main/java/io/fabric8/openshift/client/internal/OpenShiftOAuthInterceptor.java

There's a special case for RBAC review objects since these can return 200 or 201 even when unauthorized. These are set to always retry. As a consequence, I believe each SelfSubjectAccessReview will be sent twice unfortunately. This request doesn't get resent if there is no username/password or token in the OpenShiftConfig object, which should be the case in our tests, but not in practice.

The difference we're seeing between 5.4.1 and 5.6.0 is due to this PR: https://github.com/fabric8io/kubernetes-client/pull/3158. The test was failing because the OpenShiftClient found a token to use within our ~/.kube/config. If you rename that file, the test will pass.